feat(provider): add ApertoDNS Dynamic DNS provider (#1074)
qdm12/ddns-updater: docs/apertodns.md

# ApertoDNS

## Configuration

### Example

```json
{
  "settings": [
    {
      "provider": "apertodns",

>       "domain": "apertodns.com",

Reviewer: to match others

```suggestion
      "domain": "domain.com",
```

Author: Done

>       "owner": "home",

Reviewer: owner is for retrocompat only

```suggestion
```

Author: Done

      "token": "apertodns_live_xxxxxxxxxxxxx",
      "ip_version": "ipv4",
      "ipv6_suffix": ""
    }
  ]
}
```

### Compulsory parameters

> - `"provider"`: `"apertodns"`

Reviewer: nit unneeded

```suggestion
```

Author: Done

> - `"domain"`: Your ApertoDNS domain (e.g., `"apertodns.com"` or your custom domain)

Reviewer: ```suggestion
- `"domain"`: Your ApertoDNS domain, for example `a.domain.com` or `domain.com`
```

Author: Done

> - `"owner"`: The subdomain/hostname (e.g., `"home"`, `"office"`, `"@"` for root)

Reviewer: ```suggestion
```

Author: Done

- `"token"`: Your ApertoDNS API token (starts with `apertodns_live_` or `apertodns_test_`)

### Optional parameters

- `"ip_version"`: `"ipv4"` (A records), `"ipv6"` (AAAA records), or `"ipv4 or ipv6"`

> - `"ipv6_suffix"`: IPv6 suffix for EUI-64 addressing

Reviewer: ```suggestion
- `"ipv6_suffix"` is the IPv6 interface identifier suffix to use. It can be for example `0:0:0:0:72ad:8fbb:a54e:bedd/64`. If left empty, it defaults to no suffix and the raw temporary IPv6 address of the machine is used in the record updating. You might want to set this to use your permanent IPv6 address instead of your temporary IPv6 address.
```

Author: Done

> - `"base_url"`: Custom API endpoint (default: `"https://api.apertodns.com"`)

Reviewer: let's rename to `api_endpoint` instead of `base_url`

Author: Done

## Domain setup

1. Sign up at [apertodns.com](https://apertodns.com)
2. Create a domain in your dashboard
3. Generate an API token from Settings -> API Keys
4. Use the token in your configuration

## Protocol Support

This provider implements the full [ApertoDNS Protocol v1.2](https://docs.apertodns.com/protocol) with automatic fallback:

1. **Modern API** (primary): `POST /.well-known/apertodns/v1/update`
   - Bearer token authentication
   - JSON request/response

2. **Legacy DynDNS2** (fallback): `GET /nic/update`
   - Basic authentication
   - Text response ("good", "nochg")

The provider automatically uses the modern API first. If the server doesn't
support the modern endpoint (404) or has a temporary error (500), it falls
back to the legacy DynDNS2 endpoint.

**Note**: Authentication errors (invalid token, hostname not found) do NOT

> trigger fallback, as these would fail on both endpoints.

Reviewer: I appreciate the extra effort, but let's keep to use their modern API. Otherwise the user can just use the dyndns custom provider to do this.

Author: Done, the DynDNS2 fallback path is gone entirely, along with `updateLegacy` and its test. The provider now only speaks the modern endpoint.

## Modern API Endpoint

```
POST {base_url}/.well-known/apertodns/v1/update
Authorization: Bearer {token}
Content-Type: application/json

{
  "hostname": "home.apertodns.com",
  "ipv4": "93.44.241.82"
}
```

### Response (Success)

```json
{
  "success": true,
  "data": {
    "hostname": "home.apertodns.com",
    "ipv4": "93.44.241.82",
    "ipv6": null,
    "ttl": 300,
    "updated_at": "2025-01-02T12:00:00.000Z"
  }
}
```

### Response (Error)

```json
{
  "success": false,
  "error": {
    "code": "invalid_token",
    "message": "Invalid or expired token"
  }
}
```

## Legacy DynDNS2 Endpoint

```
GET {base_url}/nic/update?hostname={hostname}&myip={ip}
Authorization: Basic base64(token:{token})
```

### Response

```
good 93.44.241.82
nochg 93.44.241.82
```

## Error Codes

| Code | HTTP Status | Meaning |
|------|-------------|---------|
| `invalid_token` | 401 | Invalid or expired token |
| `unauthorized` | 401 | Missing authentication |
| `hostname_not_found` | 404 | Hostname not found |
| `invalid_ip` | 400 | Invalid IP address format |

> | `rate_limited` | 429 | Rate limit exceeded |

Reviewer: sounds like implementation details, let's remove this

```suggestion
```

Author: Done, the doc now only covers configuration.
